pkg/controller/clusters/status.go

```diff
@@ -0,0 +1,187 @@
+		return nil
+	}
+
+	if kcp == nil {
+		return fmt.Errorf("expected reference to kubeadmcontrolplane, but got nil instead ")
+	}
+
+	statusUpToDate := kcp.Status.ObservedGeneration == kcp.ObjectMeta.Generation
+	// We make sure to check that the status is up to date before using it
+	if !statusUpToDate {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.PendingUpdateReason, clusterv1.ConditionSeverityInfo, "")
+		return nil
+	}
+
+	con, err := checkKubeadmControlPlaneError(cluster, anywherev1.ControlPlaneReadyCondition, kcp, clusterv1.ReadyCondition)
+	if err != nil {
+		return err
+	}
+
+	if con != nil {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, con.Reason, con.Severity, con.Message)
+		return nil
+	}
+
+	// The control plane should be marked ready when the total number of nodes specified in the spec is
+	// equal to the ready amount in the cluster. We need to check this in two parts, since
+	// in the case of a rolling upgrade, there can be ready machines in the cluster with the old spec.
+
+	expected := cluster.Spec.ControlPlaneConfiguration.Count
+	readyReplicas := int(kcp.Status.ReadyReplicas)
+	updatedReplicas := int(kcp.Status.UpdatedReplicas)
+	totalReplicas := int(kcp.Status.Replicas)
+
+	// Get the number of outdated nodes, as long as there are some, we want to reflect that in the message
+	totalOutdated := totalReplicas - updatedReplicas
+	if totalOutdated > 0 {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.WaitingForControlPlaneReadyReason, clusterv1.ConditionSeverityInfo, "Control plane not up-to-date yet, %d rolling (%d up to date)", expected, updatedReplicas)
+		return nil
+	}
+
+	// First, we compare the total number of readyReplicas to the expected number,
+	if readyReplicas != expected {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.WaitingForControlPlaneReadyReason, clusterv1.ConditionSeverityInfo, "Control plane nodes not ready yet, %d expected (%d ready)", expected, readyReplicas)
+		return nil
+	}
+
+	conditions.MarkTrue(cluster, anywherev1.ControlPlaneReadyCondition)
+	return nil
+}
+
+// UpdateWorkersReadyCondition checks the WorkersReadyConditon condition. It checks whether each MachineDeployment is ready
+// and number of ready vs expected replicas are accounted for.
+func UpdateWorkersReadyCondition(cluster *anywherev1.Cluster, machineDeployments []clusterv1.MachineDeployment) error {
+	expected := 0
+	for _, wng := range cluster.Spec.WorkerNodeGroupConfigurations {
+		expected += *wng.Count
+	}
+
+	// The workers condition should be marked ready when the total number of nodes specified in the spec is
+	// equal to the ready amount in the cluster and there are no lingering old nodes. We need to check this in
+	// two parts, since in the case of a rolling upgrade, there can be ready machines in the cluster with the old spec.
```

Comment:
How do we check old spec? Based on observed generation field or something else?

Reply:
Or is `md.Status.UpdatedReplicas` enough for that?

Reply:
Then `md.Status.UpdatedReplicas` helps us determine how many nodes currently have the old spec. It's the total number of non-temrinated machines that have the desired template spec [according to CAPI](https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#MachineDeploymentStatus).

But checking the observed generation helps us make sure those values on the status is reflecting the status of the current desired spec first.